### PR TITLE
treewide: don't hardcode /nix/store (no rebuilds changes)

### PR DESCRIPTION
**Pull request by Artturin**
improve experience for other store locations

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/systemd/default.nix [hunk 1]
```diff
@@ -304,8 +304,8 @@ stdenv.mkDerivation (finalAttrs: {
         if dl.pkg == null then ''
           # remove the dependency on the library by replacing it with an invalid path
           for file in $(grep -lr '"${dl.name}"' src); do
-            echo "patching dlopen(\"${dl.name}\", …) in $file to an invalid store path ("/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-not-implemented/${dl.name}")…"
-            substituteInPlace "$file" --replace '"${dl.name}"' '"/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-not-implemented/${dl.name}"'
+            echo "patching dlopen(\"${dl.name}\", …) in $file to an invalid store path ("${builtins.storeDir}/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-not-implemented/${dl.name}")…"
```
Comment (wegank) [inline pkgs/os-specific/linux/systemd/default.nix:307]: Making a comment storeDir-dependent in `systemd` doesn't feel right...

File: pkgs/development/libraries/sope/default.nix [hunk 2]
```diff
@@ -35,8 +35,8 @@ gnustep.stdenv.mkDerivation rec {
 
   # Yes, this is ugly.
   preFixup = ''
-    cp -rlPa $out/nix/store/*/* $out
-    rm -rf $out/nix/store
+    cp -rlPa $out${builtins.storeDir}/*/* $out
```
Comment (wegank) [inline pkgs/development/libraries/sope/default.nix:38]: Nor here, I will rather expect a fix of the package.

File: pkgs/os-specific/linux/systemd/default.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/os-specific/linux/systemd/default.nix:307]: > Making a comment storeDir-dependent in `systemd` doesn't feel right: this mainly means that the user has to rebuild everything...

the cache already depends on storeDir so everything has to be rebuilt anyways

**Comment by Artturin** [comment]
Unnecessary for this pr but here's a useful command(s)

```
sudo mkdir -p /test-nix/var/nix && sudo chown --recursive $USER:$USER /test-nix

$(nix build --impure --expr 'with import <nixpkgs> { }; (nix.override { storeDir = "/test-nix/store"; stateDir = "/test-nix/var"; }).out' --no-link --print-out-paths)/bin/nix build ".#bash"
```

**Comment by Artturin** [comment]
Thanks @wegank

**Pull request merged**